### PR TITLE
net: lwm2m: wait for interface up in lwm2m_engine_start()

### PR DESCRIPTION
**Pull request by mike-scott**
The LwM2M subsystem should make sure that the network interface
is up before proceeding with communications.

Let's add a wait in lwm2m_engine_start() that will timeout after
60 seconds, letting the client decide if it should retry, etc.

Signed-off-by: Michael Scott <mike@foundries.io>

### CODE REVIEW

**Comment by mike-scott** [comment]
@therealprof feel free to test.  This normally doesn't affect ethernet interfaces, but always nice to make sure it doesn't break anything.

**Comment by mike-scott** [comment]
The idea behind this patch is to avoid every single user of the LwM2M subsystem having to add checks for the network interface being up.  This fixes a startup error in communications for 802.15.4, LTE-M and I think avoids occasional issues w/ bluetooth during startup.

**Comment by mike-scott** [comment]
There is an obvious long term fix that needs to be worked on where the LwM2M engine handles network UP/DOWN events after the start up.  But, this is more complicated and needs some time to work out.

File: subsys/net/lib/lwm2m/lwm2m_engine.c
Comment (jukkar) [inline subsys/net/lib/lwm2m/lwm2m_engine.c]: At some point (no need to do this for this PR), we should convert this to select correct network interface depending on destination IP address. This is very much needed if there are multiple network interfaces in the system. One such example is ethernet + VLAN support which will create multiple network interfaces in the system. For IPv4, we already have `net_if_ipv4_select_src_iface()` function, but corresponding IPv6 function is missing (I will add it soon). Also more generic version like `net_if_select_src_iface(struct sockaddr *addr)` might be useful (I will add that too).

**Review by jukkar — APPROVED** [review]

**Comment by pfalcon** [comment]
Are you aware that we have generic "net_config" subsystem whose job is to perform network configuration and bring-up, which also includes waiting for network up: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/config/Kconfig#L22 ?

> The idea behind this patch is to avoid every single user of the LwM2M subsystem having to add checks for the network interface being up.

It seems that the idea of this patch is that LwM2M subsystem, HTTP subsystem, HTTPS subsystem, COAP subsystem, DNS subsystem, MQTT subsystem - etc. - should all have their own checks for the network interface being up.

File: subsys/net/lib/lwm2m/lwm2m_engine.c
Comment (mike-scott) [reply subsys/net/lib/lwm2m/lwm2m_engine.c]: @jukkar This PR can wait till those functions are merged, and I'll rework to use them.

**Comment by mike-scott** [comment]
> Are you aware that we have generic "net_config" subsystem whose job is to perform network configuration and bring-up, which also includes waiting for network up: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/config/Kconfig#L22 ?
>
> The idea behind this patch is to avoid every single user of the LwM2M subsystem having to add checks for the network interface being up.
> 
> It seems that the idea of this patch is that LwM2M subsystem, HTTP subsystem, HTTPS subsystem, COAP subsystem, DNS subsystem, MQTT subsystem, etc. - should all have their own checks for the network interface being up.

@pfalcon What if the user overrides settings in the net_config subsystem?  Or even better, doesn't use it at all?  Are we mandating that samples use net_config?  What if a user needs to perform something in their app prior to network init?

Here, the user sets CONFIG_NET_APP_INIT_TIMEOUT to 0 (or doesn't use net_config) and then has to implement a wait for the network interface to be up later.  That check can be done incorrectly (as @jukkar pointed out in another PR), causing the net management event stack to overflow with all manner of odd results.

I felt it was best, if the LwM2M subsystem "just handles" it with the correct behavior.

Worst case: the LwM2M subsys is doing a duplicate check on the interface to make sure it's up (when the net_config later has already done so).  Is this really "bad"?

**Comment by mike-scott** [comment]
Heh, looking at how net/lib/config was split out it does seem rather hard to "disable" it.

The config options which cause problems for the LwM2M subsystem would be:
CONFIG_NET_APP_AUTO_INIT=n
or
CONFIG_NET_APP_INIT_TIMEOUT=0

Also, I'm seeing issues with offloaded network drivers.  They seem to bypass the wait for network as well.

**Comment by mike-scott** [comment]
@pfalcon No response for my later comment?

> The config options which cause problems for the LwM2M subsystem would be:
> CONFIG_NET_APP_AUTO_INIT=n
> or
> CONFIG_NET_APP_INIT_TIMEOUT=0

These effectively disable the network init / waiting for init functions of net/lib/config and there's no way around it.  How should subsystems handle a interfaces which might or might not be up?

**Comment by mike-scott** [comment]
Remember there's more to making sure that the network is up in net/lib/config.  There are the IP address settings also which are useful even when you don't want to wait for the network to be up before the application runs.

**Comment by pfalcon** [comment]
> No response for my later comment?

Yes, because the argument we have is strategical - whether to grow common subsystem or dig each own corner.

> The config options which cause problems for the LwM2M subsystem would be:
CONFIG_NET_APP_AUTO_INIT=n
or
CONFIG_NET_APP_INIT_TIMEOUT=0

For any specifics my response would be: "let's fix it!" (aka proverbial "patches welcome"). That's not my code, I'm taking it as is, and making more reusable than it was. We definitely need to pull together to make it more useful.

> Remember there's more to making sure that the network is up in net/lib/config.

So, net/lib/config is a kind of networking supervisor - it configures networking, then there's a "daemon" which manages. So far, this daemon is completely dumb - it just waits static delay and terminates. It can (should!) be made more smarter. When I prepared initial split-out, I though about names like net_daemon - but hack, as I told, there's no really a "daemon" now, so name would be confusing, or net_manager, but we already have "net management API", so it would have been confusing. Word "supervisor" escaped me until now. It's never too late to rename the stuff (especially, if it actually grows the functionality).

**Comment by pfalcon** [comment]
Otherwise, I understand the difference between proper design phase and product deadlines, so don't even try to -1 this. But, but... imagine that wakaama the lib, in its function which takes a hostname and port number waited ... for interface to be up! Even a thought is hilarious, aint't it? At most what they have is generic do-whatever-you-want callback. Because "hostname/port" and "interface is up" are two different layers, and aren't dealt together in one place in the big world.

That's btw gives us plan B if we keep adding such code to Zephyr - one sweet day we'll put aside this product deadline ridden code, and take somebody else's generic code. It may be inefficient, it may be ugly, but look which good intentions *we* have for the ugly code, we'll certainly survive somebody else's ugliness ;-).

**Comment by mike-scott** [comment]
Yes, you're correct.  Having a network interface check in a lib like Wakaama is ludicrious.. because if there's a problem they just generate an error and bail.   Then the user gets to figure it out (IE. Go check the network supervisor status, and flip some knobs).

Let's examine what that would look like in Zephyr:
User has to spend hours debugging his code to figure out the reason LwM2M subsystem is complaining.  Similar to a novice Linux user spending hours figuring out some network connection issue.

Let's also examine what a "smart" network supervisor might look like for Zephyr:
1. Each of the subsys/net/lib subsections would need to be classified as 1 of the following:
  a. network config layer which is merely a set of startup rules for the network supervisor
  b. network init / management services like DHCP: these influence how the network starts and also can handle changes while running (IE: DHCP re-issues an IP address).
  c. network helper services like DNS which act to assist other network services in some form
  d. network protocol support services like COAP which don't have any threads, but are more a library of code to be used by another entity
  e. network protocols like LwM2M, MQTT and OpenThread which require threads to handle events, etc
2. Each of the above classifications would need implementation (in different forms): DNS is spread around and pretty much a 1 off implementation, others like the network protocols with threads would require full APIs w/ callbacks to be managed generically.
3. At this point, each network entity under subsys/net/lib could expect to have the network in a state where it can be used when that service is started.
4. Applications (samples) would also need to be split so that they can perform needed actions prior to network and the main core can be executed when everything's ready.

**Comment by pfalcon** [comment]
> Let's also examine what a "smart" network supervisor might look like for Zephyr:

Designing and implementing that is definitely not easy (doing reliable networking across networking layers is much less known lore than other topics in networking). But it's fair to say that such solution could/would be reused by different services, and it's the right choice from the start to **not** hardcode it into a particular service like LwM2M. You may have other constraints which may make you wanting to do it in adhoc way. As I told, I'm not trying to -1. It's just for common "FYI" that it's not the case that nobody was thinking how to do it right ("righter").

File: subsys/net/lib/lwm2m/lwm2m_engine.c
Comment (jukkar) [reply subsys/net/lib/lwm2m/lwm2m_engine.c]: Ok, the PR is #9322, I will add you as a reviewer to that.

**Comment by jukkar** [comment]
As seen in this discussion, we are missing connection manager feature in Zephyr. This has been talked earlier in various other PRs but so far nothing has come out of it yet.

For this thing being seamless to the application wanting network connectivity, we need to separate the connectivity part into different levels. For example if application receives "network interface up" event, this is typically not enough for application as the network interface might not have yet IP address which prevents connection establishment. So we need higher level events that applications can wait in order to make sure there is possibility that connection succeeds.

I Linux side, I used to develop ConnMan connection manager. In that software we had two level events on top of ifup event:

1. "Ready" a.k.a "connected" event which is generated when IP address is set to the network interface
2. "Online" event which is generated if the manager could verify that internet connectivity is present

Something like this could be implemented in Zephyr too. The current net_config tries to answer the "Ready" case but it does not send event that application could catch but just waits until IP address is received or timeout happens. This might be too limited if the application would like more control during the device boot. Also net_config does not handle the case when network connectivity is lost and then restored.

**Comment by therealprof** [comment]
@jukkar Please don't forget the important cases: Lost network connection, received a new IP address. At the moment Zephyr really only works with permanent (static and dynamic) IP addresses and connectivity.

**Review by jukkar — CHANGES_REQUESTED** [review]
Waiting #9322 before this is to be merged.

File: subsys/net/lib/lwm2m/lwm2m_engine.c
Comment (mike-scott) [reply subsys/net/lib/lwm2m/lwm2m_engine.c]: @jukkar I tested your PR and unfortunately it doesn't work well as well as I had hoped.

The network interface needs to be "up" prior to the net_app_init_udp_client() call in lwm2m_engine_start().  There are times when we perform network access such as DNS lookup inside net_app_init_udp_client().  Of course, if we need to use DNS then I don't know if the result will be an IPv4 or IPv6 address for the destination.  And I need that destination to choose which interface to make sure is "up"... :/

Comment (pfalcon) [reply subsys/net/lib/lwm2m/lwm2m_engine.c]: >  And I need that destination to choose which interface to make sure is "up"... :/

Which shows again that working with "interfaces" in the code which is otherwise intended to work with "addresses and ports" is wrong...

Comment (jukkar) [reply subsys/net/lib/lwm2m/lwm2m_engine.c]: @mike-scott Ok, we can go with the current solution if you really need to have this functionality in lwm2m code. Eventually it would be better to remove this iface check from individual component and let system to send event when everything is ready to be connected. But this might take some time before we can have this functionality. I can start to investigate this but it will probably go to 1.14 as merge window is closing soonish.

**Review by jukkar — APPROVED** [review]

**Pull request closed without merging**
